ThatDefaultProject/Assets/ThatDefaultProject/Scripts/BehaviorTree/BehaviorTree.cs

```diff
@@ -0,0 +1,166 @@
+                _currentState = BehaviorState.Failure;
+                return _currentState;
+            }
+        }
+
+        public class BehaviorSequence : BehaviorComposite
+        {
+            public BehaviorSequence(List<Behavior> childBehaviors) : base(childBehaviors)
+            {
+            }
+
+            public override BehaviorState Execute(Blackboard blackboard)
+            {
+                foreach (var child in _childBehaviors)
+                {
+                    _currentState = child.Execute(blackboard);
+
+                    switch (_currentState)
+                    {
+                        case BehaviorState.Running:
+                        case BehaviorState.Failure:
+                            return _currentState;
+                        case BehaviorState.Success:
+                            continue;
+                    }
+                }
+                _currentState = BehaviorState.Success;
+                return _currentState;
+            }
+        }
+        #endregion
+
+        #region CONDITIONAL
+        public class BehaviorConditional : Behavior
+        {
+            private Func<Blackboard, bool> _evalFunc;
+
+            public BehaviorConditional(Func<Blackboard, bool> evalFunc)
+            {
+                _evalFunc = evalFunc;
+            }
+
+            public override BehaviorState Execute(Blackboard blackboard)
+            {
+                if (_evalFunc == null)
+                {
+                    return BehaviorState.Failure;
+                }
+                return _evalFunc(blackboard) ? BehaviorState.Success : BehaviorState.Failure;
+            }
+        }
+
+        public class BehaviorInvertConditional : Behavior
+        {
+            private Func<Blackboard, bool> _evalFunc;
+
+            public BehaviorInvertConditional(Func<Blackboard, bool> evalFunc)
+            {
+                _evalFunc = evalFunc;
+            }
+
+            public override BehaviorState Execute(Blackboard blackboard)
+            {
+                if (_evalFunc == null)
+                {
+                    return BehaviorState.Failure;
+                }
+                return _evalFunc(blackboard) ? BehaviorState.Failure : BehaviorState.Success;
+            }
+        }
+        #endregion
+
+        public class BehaviorAction : Behavior
+        {
+            private Func<Blackboard, BehaviorState> _actionFunc;
+
+            public BehaviorAction(Func<Blackboard, BehaviorState> actionFunc)
+            {
+                _actionFunc = actionFunc;
+            }
+
+            public override BehaviorState Execute(Blackboard blackboard)
+            {
+                if (_actionFunc == null)
+                {
+                    return BehaviorState.Failure;
+                }
+                _currentState = _actionFunc(blackboard);
+                return _currentState;
+            }
+        }
+
+        public interface IDecisionMaking
```

Comment:
Why is an interface needed if the only class deriving from it is the BehaviorTree?